nullaway/src/test/resources/com/uber/nullaway/testdata/NullAwayStreamSupportNegativeCases.java

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright (c) 2019 Uber Technologies, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+package com.uber.nullaway.testdata;
+
+import com.google.common.collect.ImmutableList;
+import java.util.function.*;
+import java.util.stream.*;
+import javax.annotation.Nullable;
+
+public class NullAwayStreamSupportNegativeCases {
+
+  static class NullableContainer<T> {
+    @Nullable private T ref;
+
+    public NullableContainer() {
+      ref = null;
+    }
+
+    @Nullable
+    public T get() {
+      return ref;
+    }
+
+    public void set(T o) {
+      ref = o;
+    }
+  }
+
+  private static boolean perhaps() {
+    return Math.random() > 0.5;
+  }
+
+  private Stream<Integer> filterThenMap(Stream<String> stream) {
+    return stream
+        .filter(
+            new Predicate<String>() {
+              @Override
+              public boolean test(String s) {
+                return s != null;
```

Comment:
Ok, how about this same example but with `return true;` here? Do we get a FN? Would we have had it before this change?

And what if there is no filter but just directly the `map`? What if we use lambdas?

I think this might be a FN we _always_ had, by virtue of `Function` being unannotated and our (unsound) optimistic assumptions. But it's worth double checking.

If the hole was there to begin with and the other nits are addressed, then this PR looks pretty good to me! (Thanks for refactoring one of the gnarliest parts of NullAway 😉 )

Reply:
As @msridhar mentioned, I am pretty sure we were always missing the cases above, but if you have a moment to double check (by commenting and uncommenting the line in `Handlers` and seeing the FN occurs either way), that'd be great.

Reply:
> As @msridhar mentioned, I am pretty sure we were always missing the cases above, but if you have a moment to double check (by locally adding that test case, then commenting and uncommenting the line in `Handlers` and seeing the FN occurs either way), that'd be great.

Sorry. What is FN ?

Reply:
FN is False Negative

Reply:
After this pr this code gives a FN:
```
private Stream<Integer> filterWithIfThenMapNullableContainerNullableOnSomeBranch(
      Stream<NullableContainer<String>> stream) {
    return stream
        .filter(
            new Predicate<NullableContainer<String>>() {
              @Override
              public boolean test(NullableContainer<String> container) {
                return true;
              }
            })
        .map(
            new Function<NullableContainer<String>, Integer>() {
              @Override
              public Integer apply(NullableContainer<String> c) {
                // BUG: Diagnostic contains: dereferenced expression
                return c.get().length();
              }
            });
  }
```
Before the fix it failed with 
> [NullAway] dereferenced expression c.get() is @Nullable
>                    return c.get().length();

So we have here a FN

Reply:
This case is a FN before and after the fix

```
private Stream<Integer> filterThenMap(Stream<String> stream) {
    return stream
            .filter(
                    new Predicate<String>() {
                      @Override
                      public boolean test(String s) {
                        return true;
                      }
                    })
            .map(
                    new Function<String, Integer>() {
                      @Override
                      public Integer apply(String s) {
                        return s.length();
                      }
                    });
  }
```

This case fails before and after fix
```
 private Stream<Integer> filterThenMap(Stream<NullableContainer<String>> stream) {
    return stream
            .filter(
                    new Predicate<NullableContainer<String>>() {
                      @Override
                      public boolean test(NullableContainer<String> s) {
                        return true;
                      }
                    })
            .map(
                    new Function<NullableContainer<String>, Integer>() {
                      @Override
                      public Integer apply(NullableContainer<String> s) {
                        return s.get().length();
                      }
                    });
  }
```

so it looks we are good to go

Reply:
This test case is also a FN before and after our fix

```
private Stream<Integer> filterThenMap(Stream<String> stream) {
    return stream
            .map(s->s == null? s: null)
            .map(
                    new Function<String, Integer>() {
                      @Override
                      public Integer apply(String s) {
                        return s.length();
                      }
                    });
  }
```

Reply:
Ok, that's what I expected. Sorry for the trouble and thanks for being so through! :)